### PR TITLE
types improvements

### PR DESCRIPTION
**Pull request by Wroud**
change actionTypes type from `Map<true>` to `Array<string>`
change some types, add `Dictionary<T>` and replace `Record<T>` with them
make `createStore` first argument optional
replace `throwIf` with `throwError`

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -2,8 +2,8 @@ console.warn('REAtom still work in progress, do not use it in production')
 
 type NodeId = string
 
-type ActionType = string
-type ActionTypesDictionary = Record<ActionType, true>
+type Dictionary<T> = { [key: string]: T }
+type ActionTypesDictionary = Record<NodeId, true>
```
Comment (artalar) [inline src/index.ts]: ` type ActionTypesDictionary = Record<ActionType, true>` was right

File: src/index.ts [hunk 2]
```diff
@@ -2,8 +2,8 @@ console.warn('REAtom still work in progress, do not use it in production')
 
 type NodeId = string
 
-type ActionType = string
-type ActionTypesDictionary = Record<ActionType, true>
+type Dictionary<T> = { [key: string]: T }
```
Comment (artalar) [inline src/index.ts]: ` type Dictionary<T> = { [key: string]: T | undefined }`

File: src/index.ts (hunk 1 above)
Comment (artalar) [reply src/index.ts]: But you may replace Record by Dictionary

Comment (Wroud) [reply src/index.ts]: mean like `type ActionTypesDictionary = Dictionary<bool>`?

File: src/index.ts [hunk 3]
```diff
@@ -46,7 +46,7 @@ let id = 0
 function nameToId(name: unknown) {
   return Array.isArray(name)
     ? safetyStr(name[0], 'name')
-    : safetyStr(name, 'name') + ' #' + ++id
+    : `${safetyStr(name, 'name')} #${++id}`
```
Comment (belozer) [inline src/index.ts]: don't. It's a very slow operation.

File: src/index.ts [hunk 4]
```diff
@@ -155,7 +155,7 @@ export function declareAtom<State>(
 
   throwIf(
     initialState === undefined,
-    'Atom "' + atomId + '". Initial state can\'t be undefined',
+    `Atom "${atomId}". Initial state can't be undefined`,
```
Comment (belozer) [inline src/index.ts]: don't. It's a very slow operation.

File: src/index.ts [hunk 5]
```diff
@@ -220,13 +220,7 @@ export function declareAtom<State>(
...
         throwIf(
           atomStateNew === undefined,
-          'Invalid state. Reducer №' +
-          position +
-          ' in ' +
-          '"' +
-          atomId +
-          '"' +
-          ' atom returns undefined',
+          `Invalid state. Reducer № ${position} in "${atomId}" atom returns undefined`,
```
Comment (belozer) [inline src/index.ts]: don't. It's a very slow operation.

File: src/index.ts [hunk 6]
```diff
@@ -336,7 +330,7 @@ export function combine(name: any, shape: any) {
   if (arguments.length === 1) {
     // @ts-ignore
     shape = name
-    name = '{' + (keys = Object.keys(shape)).join() + '}'
+    name = `{${(keys = Object.keys(shape)).join()}}`
```
Comment (belozer) [inline src/index.ts]: don't. It's a very slow operation.

File: src/index.ts (hunk 4 above)
Comment (Wroud) [reply src/index.ts]: its exception....

Comment (belozer) [reply src/index.ts]: It does not matter :) it is Better to do work with strings in the same style.

Comment (Wroud) [reply src/index.ts]: but code will run anyway so better implement string formatting in `throwIf`  if we want better perfomace

Comment (belozer) [reply src/index.ts]: But for exceptions... Maybe we should talk about it.

Comment (belozer) [reply src/index.ts]: About better perfomance. See https://gist.github.com/c564436f39a8a1557b5969f405ccb492

Comment (belozer) [reply src/index.ts]: rather it is http://jsbench.github.io/#c564436f39a8a1557b5969f405ccb492

File: src/index.ts [hunk 7]
```diff
@@ -524,21 +516,30 @@ function callFromList(list: Function[], arg: any, i = -1) {
   while (++i < list.length) list[i](arg)
 }
 
+function formatString(message: string, ...args: any[]) {
```
Comment (belozer) [inline src/index.ts]: Loss in performance and extra weight to the package.

File: package.json [hunk 8]
```diff
@@ -18,11 +18,14 @@
     "@babel/preset-env": "^7.5.5",
     "@babel/preset-typescript": "^7.3.3",
     "@types/jest": "^24.0.18",
+    "@types/react": "^16.8.0",
```
Comment (belozer) [inline package.json]: it's not necessary here.

Comment (belozer) [reply package.json]: react and redux too

Comment (belozer) [reply package.json]: your need run `npm i` from `packages/packageName/`. Adding them here is not the solution.

File: lerna.json [hunk 9]
```diff
@@ -2,7 +2,7 @@
   "packages": [
     "packages/*"
   ],
-  "hoist": true,
+  "hoist": false,
```
Comment (belozer) [inline lerna.json]: you're working with true, right?

Comment (Wroud) [reply lerna.json]: yep

Comment (belozer) [reply lerna.json]: so it's not necessary here

Comment (Wroud) [reply lerna.json]: its kind are bug, this change was mark outdated

Comment (belozer) [reply lerna.json]: ohhh.... Sorry

**Pull request closed without merging**